File: Heap/Find_K_Pairs_with_Smallest_Sums.py
#Problem Link:- https://leetcode.com/problems/find-k-pairs-with-smallest-sums/

"""You are given two integer arrays nums1 and nums2 sorted in ascending order and an 
integer k.Define a pair (u,v) which consists of one element from the first array and 
one element from the second array.Find the k pairs (u1,v1),(u2,v2) ...(uk,vk) with 
the smallest sums."""

class Solution(object):
    def kSmallestPairs(self, nums1, nums2, k):
        import heapq
        heap = []
        n = len(nums1)
        m = len(nums2)
        for i in range(n):
            for j in range(m):
                heappush(heap,(nums1[i] + nums2[j],(nums1[i],nums2[j])))
        ans = []
        for i in range(min(n*m,k)):
            sm,pair = heappop(heap)
            ans.append(pair)
        return ans

File: Heap/FindKPairsWithSmallestSums.cpp
//Problem Link:- https://leetcode.com/problems/find-k-pairs-with-smallest-sums/

#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<vector<int>> kSmallestPairs(vector<int>& nums1, vector<int>& nums2, int k) {
        priority_queue<pair<int,vector<int>>> pq;
        int n = nums1.size();
        int m = nums2.size();
        for(int i=0;i<=min(n-1,k);i++){
            for(int j=0;j<=min(m-1,k);j++){
                int sm = nums1[i] + nums2[j];
                if (pq.size() == k){
                    if (sm >= pq.top().first){
                        continue;
                    }
                    else{
                        pq.pop();
                        pq.push({sm,{nums1[i],nums2[j]}});
                    }
                }
                else{
                    pq.push({sm,{nums1[i],nums2[j]}});
                }
            }
        }
        vector<vector<int>> result;
        while (pq.size() > 0){
            auto node = pq.top();
            pq.pop();
            result.push_back(node.second);
        }
        return result;
    }
};